californium-core/src/main/java/org/eclipse/californium/core/network/Matcher.java

```diff
@@ -145,33 +159,75 @@ public synchronized void setExecutor(final ScheduledExecutorService executor) {
 		// health status runnable is not migrated at the moment
 	}
 
-	public void sendRequest(Exchange exchange, Request request) {
+	public void sendRequest(final Exchange exchange, final Request request) {
 
-		// ensure MID is set
+		// request MID is from the local name space -- use blank address
+		final KeyMID idByMID = assertOutboundRequestHasMessageId(request);
+
+		// ensure Token is set
+		final KeyToken idByToken = assertOutboundRequestHasToken(exchange, request);
+
+		exchange.setObserver(exchangeObserver);
+		exchangesByMID.put(idByMID, exchange);
+		exchangesByToken.put(idByToken, exchange);
+		addToObservationStore(exchange, request);
```

Comment:
In cluster mode, we need this coordinated NotificationOrderer.
When not in cluster mode, the current NotificationOrderer should have the same interface as the cluster NotificationOrderer and be called at the same point.